Collection of small fixes for splash project

grondo opened the pull request:
Here's a collection of small fixes related to the splash use case for consideration.

This PR includes:
 * General reduction in logging including removal of unnecessary log messages, reduction of some "info" level log messaages to debug, and fix for #1439 to log only from rank 0 instead of all ranks
 * Fix for crash in `flux-wreck status` when `exit_status` is not in kvs yet (#1437)
 * Increase NOFILE_LIMIT in wrexecd (#1441)
 * Error checking during task setup in wrexecd (#1441)
 * Add line buffering to output files with `flux submit/wreckrun -O file`
 * Check for errors from kz stream in output.lua and log errors  (#1427)
 * Terminate io "flush" loop early on encountering any error from `f:reactor ("once")` 
 * New command `flux-wreck dumplog ID` to dump the "logstream" error log for a job
 * Add new KZ_FLAGS_NOFOLLOW flag as suggested by @garlick (#1420)
 * Add `-n, --no-follow` option to `flux-wreck attach` to use KZ_FLAGS_NOFOLLOW (#1420)

Might need to add more testing (e.g. of `flux wreck attach -n`) but since there is lots of little things in here I wanted to get an early review.

garlick commented:
Restarted one builder that had stopped due to #1077
```
scripts/waitfile works after 1s

expecting success: 
	flux start ${ARGS} -s2 --bootstrap=selfpmi bash -c "nohup flux event sub hb &"

lt-flux-broker: module 'job' was not cleanly shutdown
lt-flux-broker: module 'barrier' was not cleanly shutdown
lt-flux-broker: module 'resource-hwloc' was not cleanly shutdown
lt-flux-broker: module 'aggregator' was not cleanly shutdown
lt-flux-broker: module 'kvs' was not cleanly shutdown
flux-start: 0 (pid 11626) Killed
flux-event: flux_event_subscribe: Success
not ok 53 - instance can stop cleanly with subscribers (#1025)
```

garlick commented on src/bindings/lua/wreck.lua:
```diff
@@ -470,13 +475,24 @@ function wreck.logstream (arg)
-                io.stderr:write (r.."\n")
-            end
-        }
+        local iow
```
> The problem with using the reactor is that log streams aren't properly
terminated with EOF at job end, so reactor based read of the kz
files doesn't know when it is done...

This is fine, but doesn't KZ_FLAGS_NOFOLLOW allow this to work with the reactor?

grondo replied:
Yeah, this commit was done before the KZ_FLAGS_NOFOLLOW work, so I should probably take a second look. I'd like to try and see if the reactor would indeed exit normally with a set of kz streams open with KZ_FLAGS_NOFOLLOW. The way it is done now might be more predictable, but at the very least the commit message is misleading, sorry.

grondo commented:
Ok, I switched the logstream "oneshot" to use KZ_FLAGS_NOFOLLOW instead of using kz_open instead of iowatchers. 

I also added @trws's fix for `-t, --tasks-per-node` along with a new test for that functionality.

Finally, a fix for #1453, along with a test.

garlick commented:
Nice - will merge once travis finishes.

grondo commented:
Broke the pmi tests with the `-t, --tasks-per-node` fix. Now that wreck.tasks_per_node was always initialized that broke the "fake" R_lite generation when wreckrun is run without a scheduler (however, only with specific combinations of arguments).

Added a fix and rebased on current master, we'll see if this passes Travis.

trws commented:
Cool, thanks for putting these together so fast!

On 12 Apr 2018, at 14:47, Mark Grondona wrote:

> Broke the pmi tests with the `-t, --tasks-per-node` fix. Now that 
> wreck.tasks_per_node was always initialized that broke the "fake" 
> R_lite generation when wreckrun is run without a scheduler (however, 
> only with specific combinations of arguments).
>
> Added a fix and rebased on current master, we'll see if this passes 
> Travis.
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/1450#issuecomment-380954440

grondo commented:
Oops forgot to add a test for the KZ_FLAGS_NOFOLLOW usage in attach/dumplog

grondo commented:
Thanks!
